### PR TITLE
Add time to kafka event

### PR DESCRIPTION
**Pull request by lnahiro**
- Add timing on Kafka event
- Add callback for produce message (response & error)
- Add event_pool_timeout configuration

### CODE REVIEW

File: Helper/ExceptionFormatterTrait.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+namespace M6Web\Bundle\KafkaBundle\Helper;
+
+/**
+ * Trait ExceptionFormatterTrait
+ * @package M6Web\Bundle\KafkaBundle
+ */
+trait ExceptionFormatterTrait
+{
+    protected $errorMapping = [
+        RD_KAFKA_RESP_ERR__BAD_MSG => 'Bad message format',
```
Comment (NastasiaSaby) [inline Helper/ExceptionFormatterTrait.php]: 😊

File: Helper/NotifyEventTrait.php [hunk 2]
```diff
@@ -33,16 +38,60 @@ public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
...
     /**
-     * @param string $origin origin (producer or consumer)
-     *
-     * @return void
+     * @param string $origin
+     * @param string $key
+     */
+    public function prepareEvent(string $origin, string $key)
+    {
+        $event = new KafkaEvent($origin);
```
Comment (NastasiaSaby) [inline Helper/NotifyEventTrait.php]: Is it not a pity to do that if we don't want any event?

File: Event/KafkaEvent.php [hunk 3]
```diff
@@ -32,4 +60,98 @@ public function getOrigin(): string
...
+    {
+        $this->executionStart = microtime(true);
+
+        return $this;
+    }
+
+    /**
+     * @return KafkaEvent
+     */
+    public function setExecutionStop(): KafkaEvent
```
Comment (NastasiaSaby) [inline Event/KafkaEvent.php]: Shouldn't we call it "setExecutionTime" instead?

File: Event/KafkaEvent.php [hunk 4]
```diff
@@ -32,4 +60,98 @@ public function getOrigin(): string
...
+    {
+        return $this->executionTime;
+    }
+
+    /**
+     * Return execution time in milliseconds
+     *
+     * @return float
+     */
+    public function getTiming(): float
```
Comment (NastasiaSaby) [inline Event/KafkaEvent.php:153]: Shouldn't we call it "getExecutionTime" instead?

File: Helper/NotifyEventTrait.php [hunk 5]
```diff
@@ -33,16 +38,81 @@ public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
...
-                ->dispatch('kafka.event', new KafkaEvent($origin));
+                ->dispatch(KafkaEvent::EVENT_NAME, $event);
+        }
+    }
+
+    /**
+     * @param string $key
+     * @param int    $errorCode
+     */
+    public function notifyResponseErrorEvent(string $key, int $errorCode)
```
Comment (NastasiaSaby) [inline Helper/NotifyEventTrait.php:73]: Shouldn't we call it "notifyErrorFromConsumerEvent" cause it deals only with consumers?

File: Helper/NotifyEventTrait.php [hunk 6]
```diff
@@ -33,16 +38,81 @@ public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
...
+                ->dispatch(KafkaEvent::EVENT_ERROR_NAME, $event);
+        }
+    }
+
+    /**
+     * @param string $origin
+     * @param int    $errorCode
+     * @param string $reason
+     */
+    public function notifyErrorEvent(string $origin, int $errorCode, string $reason)
```
Comment (NastasiaSaby) [inline Helper/NotifyEventTrait.php:92]: Shouldn't we call it "notifyErrorFromProducerEvent" cause it deals only with producers?

**Review by NastasiaSaby — APPROVED** [review]

File: Helper/NotifyEventTrait.php (hunk 5 above)
Comment (lnahiro) [reply Helper/NotifyEventTrait.php:73]: i just changed, so now its used also with producer

File: Helper/NotifyEventTrait.php (hunk 6 above)
Comment (lnahiro) [reply Helper/NotifyEventTrait.php:92]: there is the same callback with consumer, but we dont use it for the moment.

File: Event/KafkaEvent.php (hunk 4 above)
Comment (lnahiro) [reply Event/KafkaEvent.php:153]: this allow to use statsd event directly and dont configure a custom method.
there is already an other method getExecutionTime used in this method.

File: Event/KafkaEvent.php (hunk 3 above)
Comment (lnahiro) [reply Event/KafkaEvent.php]: There is a calcul, isn't set the execution time.
Maybe remove the set for executionstop

**Review by Fabex — APPROVED** [review]

File: Factory/ProducerFactory.php [hunk 7]
```diff
@@ -20,12 +20,17 @@ public function get(string $producerClass, array $producerData): ProducerManager
...
+        $this->configuration->setDrMsgCb([$producerManager, 'produceResponse']);
+        $this->configuration->setErrorCb([$producerManager, 'produceError']);
+
         $producer = new $producerClass($this->configuration);
 
         $producerManager->setProducer($producer);
         $producerManager
             ->addBrokers(implode(',', $producerData['brokers']))
             ->setLogLevel($producerData['log_level'])
+            ->setEventsPoolTimeout($producerData['events_poll_timeout'])
```
Comment (Oliboy50) [inline Factory/ProducerFactory.php]: you call it `setEventsPoolTimeout` but your param is called `events_poll_timeout` (pool > poll)

could you harmonize it?

File: Manager/ProducerManager.php [hunk 8]
```diff
@@ -40,6 +42,12 @@ class ProducerManager
     protected $brokers;
 
     /**
+     * event pool timeout in ms
```
Comment (Oliboy50) [inline Manager/ProducerManager.php]: still `pool` here (and maybe it's the good word to use)

File: README.md [hunk 9]
```diff
@@ -60,6 +60,7 @@ m6_web_kafka:
                - '127.0.0.1'
                - '10.05.05.19'
            log_level: 3
+           event_pool_timeout: 2000 #ms
```
Comment (Oliboy50) [inline README.md]: still `pool` here (and maybe it's the good word to use)

**Comment by Oliboy50** [comment]
@lnahiro do we talk about [poll](http://dictionnaire.reverso.net/anglais-francais/poll) or [pool](http://dictionnaire.reverso.net/anglais-francais/pool)?
I'd say we're talking about `pool` so you should rename everything again 😛

**Comment by lnahiro** [comment]
@Oliboy50 no, its poll and not pool: https://arnaud-lb.github.io/php-rdkafka/phpdoc/rdkafka.poll.html

File: Event/KafkaEvent.php (hunk 4 above)
Comment (fabdsp) [reply Event/KafkaEvent.php:153]: WDYT of getExecutionTimeMs ?

**Review by fabdsp — APPROVED** [review]

File: Event/KafkaEvent.php (hunk 4 above)
Comment (NastasiaSaby) [reply Event/KafkaEvent.php:153]: I'm ok

**Pull request merged**